src/cpp/src/storage/checkpointer.cpp

```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
     config_ = config;
 }
 
-void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+void Checkpointer::create_checkpoint(string checkpoint_dir, CheckpointMeta checkpoint_meta, int epochs) {
+    string tmp_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "_tmp/";
```

Comment:
Why create a tmp directory and not just put the files into the checkpoint directory directly?

Reply:
a bit far-fetched. we could crash in the middle of the process. if so, we wouldn't want to use the half-written checkpoint. hence went with the rename option which I believe is atomic. so if there exists a directory of the form `checkpoint_x`, we can be sure that the checkpoint is valid.